kubespawner/spawner.py

```diff
@@ -53,17 +60,17 @@ class PodReflector(NamespacedResourceReflector):
         'component': 'singleuser-server',
     }
 
-    list_method_name = 'list_namespaced_pod'
+    list_method_name = 'list_pod_for_all_namespaces'
 
     @property
     def pods(self):
         return self.resources
 
 
-class EventReflector(NamespacedResourceReflector):
+class EventReflector(ResourceReflector):
     kind = 'events'
 
-    list_method_name = 'list_namespaced_event'
+    list_method_name = 'list_event_for_all_namespaces'
```

Comment:
The list method should depend on whether a namespace is set, right?

Reply:
There are two options here. 
- Always search globally without the namespace scope (which is the current implementation and `might` have a small performance penalty compared to looking up in a namespace)
- Conditionally use different APIs depending on how namespaces are being used (this would be similar to what was implemented in #218 , and based on the current implementation (as we always have a namespace), it will need a bunch of subclasses/ifs and I felt the code was a little convoluted) but I could go back to revisit that if you would prefer that.

Reply:
Thanks for working on this Luciano! I've read it through but not tested it.

Is it correct that not using this feature, due to this part will currently still force the privilege escalation of the pod running KubeSpawner to have ClusterRole permissions?

This may become a deal breaker for various users of KubeSpawner and Z2JH that relies on KubeSpawner. I know clusterrole wide permission can be a big ask for various administrators in general. For those that this isn't an issue, it may also require additional labor.

The more I think about it, the more afraid I'd be to click merge on something that require privilege escalation no matter what instead of making it something one opts into if needing to use a feature that requires it.

Reply:
One would only require to use the cluster-wide roles if kubespawner is creating the per-user namespaces. Otherwise, the pods would be running on the already created 'hub' namespace and the regular `role/rolebinding` would be sufficient. 

One aspect is that the current implementation does try to create the 'user namespace' if the new function is enabled which will fail if the `cluster role` is not available. What I could do is to enhance the implementation to first check if the namespace is already present before trying to create, this would enable `bring your own namespace` where the admins would previously create the namespaces for the set of users without requiring the additional `cluster roles`

Reply:
I could also provide a quick table (validated with some tests), with what works with `role` versus `cluster role`.

Reply:
> One would only require to use the cluster-wide roles if kubespawner is creating the per-user namespaces. Otherwise, the pods would be running on the already created 'hub' namespace and the regular role/rolebinding would be sufficient.

Hmmm, but right now we are are invoking a different method on the `kubernetes` python package, that refers to listing all pods and all events etc rather than those in the namespace of kubespawner / hub itself right? I don't think we would be permitted to do that no matter if we create pods only within our current namespace.

This is the role attached to the hub service account currently through a RoleBinding, that is in use by the hub pod, that kubespawner runs within. It is not going to permit anything outside the namespace, so I wonder even if you are allowed to list clusterroles (`kubectl get clusterrole`) or pods in all namespaces.

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  creationTimestamp: 2019-04-01T22:14:42Z
  labels:
    app: jupyterhub
    chart: jupyterhub-0.9-61cf357
    component: hub
    heritage: Tiller
    release: jhub
  name: hub
  namespace: jhub
  resourceVersion: "714609"
  selfLink: /apis/rbac.authorization.k8s.io/v1/namespaces/jhub/roles/hub
  uid: 8c77a641-54cb-11e9-9277-42010aa600ec
rules:
- apiGroups:
  - ""
  resources:
  - pods
  - persistentvolumeclaims
  verbs:
  - get
  - watch
  - list
  - create
  - delete
- apiGroups:
  - ""
  resources:
  - events
  verbs:
  - get
  - watch
  - list
```

I figure we need the kubespawner/hub pod to run with additional permissions granted to it by default if we invoke a kubernetes api call to get stuff from all namespaces as default. When we set the method to `list_event_for_all_namespaces` instead of `list_namespaced_event`, it is my understanding that the `kubernetes` python package that helps us make kubernetes API calls like those run by in `kubectl` the CLI will demand it.

If you got this code working on your deployment, perhaps it is due to having a custom service account on the hub pod with powerful permissions?

Reply:
@consideRatio Let me revisit this and make sure I document what works with `Role` versus `Cluster Role` as it has been a couple of weeks.

Reply:
@lresende thanks picking up the ball with this effort!   We've been running #218 in production since last September on top of z2jh, with the following ClusterRole and ClusterRoleBindings in place:   

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: hub
rules:
- apiGroups:
  - ""
  resources:
  - pods
  - events
  - persistentvolumeclaims
  verbs:
  - get
  - watch
  - list
  - delete
  - create
```

and

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: hub
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: hub
subjects:
- kind: ServiceAccount
  name: hub
  namespace: jupyterhub
```